vassal-app/src/main/java/VASSAL/configure/ConfigureTree.java

```diff
@@ -1345,86 +1517,151 @@ public SearchAction(ConfigureTree configureTree, SearchParameters searchParamete
-
-      final JButton find = new JButton(Resources.getString("Editor.search_next"));
-      find.addActionListener(new ActionListener() {
-        @Override
-        public void actionPerformed(ActionEvent e) {
+      JDialog d = configureTree.getSearchDialog();
+      JTextField search;
+      if (d != null) {
+        search = configureTree.getSearchField();
+      }
+      else {
+        d = new JDialog((Frame) SwingUtilities.getAncestorOfClass(Frame.class, configureTree), false);
+        configureTree.setSearchDialog(d);
+
+        d.setTitle(configureTree.getSearchCmd());
+
+        final JLabel searchLabel = new JLabel("String to find: ");
+        search = new JTextField(searchParameters.getSearchString(), 32);
+        configureTree.setSearchField(search);
+        search.select(0, searchParameters.getSearchString().length()); // Pre-select all the search text when opening the dialog
+        searchLabel.setLabelFor(search);
+
+        final JCheckBox sensitive = new JCheckBox(Resources.getString("Editor.search_case"), searchParameters.isMatchCase());
+        final JCheckBox advanced  = new JCheckBox(Resources.getString("Editor.search_advanced"), searchParameters.isMatchAdvanced());
+
+        final JCheckBox names = new JCheckBox(Resources.getString("Editor.search_names"), searchParameters.isMatchNames());
+        final JCheckBox types = new JCheckBox(Resources.getString("Editor.search_types"), searchParameters.isMatchTypes());
+
+        final JCheckBox traits = new JCheckBox(Resources.getString("Editor.search_traits"), searchParameters.isMatchTraits());
+        final JCheckBox expressions = new JCheckBox(Resources.getString("Editor.search_expressions"), searchParameters.isMatchExpressions());
+        final JCheckBox properties = new JCheckBox(Resources.getString("Editor.search_properties"), searchParameters.isMatchProperties());
+
+        final JCheckBox keys = new JCheckBox(Resources.getString("Editor.search_keys"), searchParameters.isMatchKeys());
+        final JCheckBox menus = new JCheckBox(Resources.getString("Editor.search_menus"), searchParameters.isMatchMenus());
+        final JCheckBox messages = new JCheckBox(Resources.getString("Editor.search_messages"), searchParameters.isMatchMessages());
+
+        advanced.addChangeListener(l -> {
+          names.setVisible(advanced.isSelected());
+          types.setVisible(advanced.isSelected());
+          traits.setVisible(advanced.isSelected());
+          expressions.setVisible(advanced.isSelected());
+          properties.setVisible(advanced.isSelected());
+          keys.setVisible(advanced.isSelected());
+          menus.setVisible(advanced.isSelected());
+          messages.setVisible(advanced.isSelected());
+          configureTree.getSearchDialog().pack();
+        });
+
+        // There's probably some easy way to just "fire the change listener" in the lambda above instead of repeating
+        // this here. If you know how, feel free to put it in.
```

Comment:
You could put all of this into a method, and then call the method both here and from the lambda.

Reply:
I thought about that but then I'd have to make all those JCheckboxes members instead of just local, and that bothered me more than just repeating the set-them code twice in a row right next to each other where you can at least see exactly what's going on all in one place. :(